pkg/resource/deploy/step_generator.go

```diff
@@ -585,8 +596,13 @@ func (sg *stepGenerator) generateStepsFromDiff(
 			//
 			// Note that if we're performing a targeted replace, we already have the correct inputs.
 			if prov != nil && !sg.isTargetedReplace(urn) {
+				// Increment the sequence number (if it's known) before calling check so we get a new autoname
+				if new.SequenceNumber != 0 {
+					new.SequenceNumber++
+				}
+
 				var failures []plugin.CheckFailure
-				inputs, failures, err = prov.Check(urn, nil, goal.Properties, allowUnknowns)
+				inputs, failures, err = prov.Check(urn, nil, goal.Properties, allowUnknowns, new.SequenceNumber)
```

Comment:
Consider the code below. Planner may be cascading along the dependency graph to delete and re-create some other resources. I am wondering if sequencenumber for these resources would auto-increment, and if so, how. It sounds like it should? Since it's an indirect replacement.

```
NewDeleteReplacementStep(sg.deployment, old, true),
NewReplaceStep(sg.deployment, old, new, diff.ReplaceKeys, diff.ChangedKeys, diff.DetailedDiff, false),
NewCreateReplacementStep(
```

Reply:
Yes it should increment and I think that will get caught by the step generator seeing that 'recreating' is true for those resources.

Reply:
Yes it would seem so from:

```
// Mark the condemned resource as deleted. We won't know until later in the deployment whether
// or not we're going to be replacing this resource.
sg.deletes[dependentResource.URN] = true
```

and this definition of recreating

```
	// We may be re-creating this resource if it got deleted earlier in the execution of this deployment.
	_, recreating := sg.deletes[urn]

```

However I do not fully grok the state machine here wrt sg.deletes and ordering.

It would seem that we'd need some sort of circularity here that might be absent, that is:

- we handle seq-num in generateSteps
- generateSteps decides to do some extra replacements based on  deleteBeforeReplace and dep graph
- in that case those replacements need seq-num handling so these dep resources need to undergo generateSteps

But it would seem that generateSteps does not recur, in fact it would seem it is called once per resource in the order as RegisterResource calls arrive. 

Let me run a quick experiment to clarify further.